File: 16.3-sum-closest.py
#
# @lc app=leetcode id=16 lang=python3
#
# [16] 3Sum Closest
#

# Time:  O(n^2)
# Space: O(1)

from typing import List


# @lc code=start
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        min_diff, closet = float("inf"), float("inf")
        nums.sort()
        for i in range(len(nums) - 2):
            left, right = i + 1, len(nums) - 1
            while left < right:
                sum = nums[i] + nums[left] + nums[right]
                diff = abs(sum - target)
                if diff < min_diff:
                    min_diff = diff
                    closet = sum
                if sum < target:
                    left += 1
                else:
                    right -= 1
        return closet


# @lc code=end

File: 16.3-sum-closest.js
/*
 * @lc app=leetcode id=16 lang=javascript
 *
 * [16] 3Sum Closest
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function (nums, target) {
  nums.sort((a, b) => a - b);
  let closet = Infinity;
  for (let i = 0; i < nums.length - 2; ++i) {
    let j = i + 1;
    let k = nums.length - 1;
    while (j < k) {
      const sum = nums[i] + nums[j] + nums[k];
      if (sum === target) {
        return target;
      }
      if (Math.abs(target - sum) < Math.abs(target - closet)) {
        closet = sum;
      }
      sum < target ? ++j : --k;
    }
  }
  return closet;
};
// @lc code=end
